Subfolders and variables/exceptions support

RamonEsteveCuevas opened the pull request:
Added support for using subfolders which also contains valid PL/SQL files. Currently it will always search in subfolders but this can be extended by using the \* sign for instance. 
In the company I work we also define variables and exceptions in the pks, these were not generated. I have created support to comment these variables and print them in the template (package.md). These variables can be commented by using the @excep (exceptions) and @var tags (variables).

martindsouza commented:
@RamonEsteveCuevas thanks for contributing to the project. I looked at the diff and having issues following changes to `dox.js` Did you change the tab structure or something since a lot of the lines are the same but marked as changed?

As for `@var` and `@excep` tags I get why they're in there. For exceptions I think we'll list as `@exception`.

Before accepting can you please do the following:
- [ ] Resubmit or clean up code so the diffs work properly. I think this may have to do with tabs in your code (we use 2 spaces for tabs)
- [ ] Provide some a sample `pks` file so I can see how you document `var`s and `exception`s

On my end once committed the documentation and examples will need to be updated. I have created #50 for this.

RamonEsteveCuevas commented:
@martindsouza Thanks for the positive response and feedback! I have changed the files so they use indentation of 2 instead of 4 now. Also changed the tag from @excep to @exception. The branch now contains a config.json which I used to generate the documentation and it also contains sample pks files (src, src/subfolder).

martindsouza commented on lib/dox.js:
```diff
@@ -31,19 +31,19 @@ dox.contextPatternMatchers = [
         // Comments after constants may still be in the array.
         // As such only allow definitions of constants
-        if (/^\s*\w+\s+constant\s+.+:=.+$/gi.test(constant)){
+        if (/^\s*\w+\s+constant\s+.+(:=|default).+$/gi.test(constant)){
```
Why a `default` for a constant?

RamonEsteveCuevas replied:
While writing our program we use the default operator as assignment target in the declaration sections of our code. This is still valid PL/SQL and the framework wasn't picking these variables up. As seen in the oracle documentation: https://docs.oracle.com/cd/B28359_01/appdev.111/b28370/fundamentals.htm#CHDFADII

martindsouza commented on lib/dox.js:
```diff
@@ -56,6 +56,84 @@ dox.contextPatternMatchers = [
+  //Variables
+  function (str) {
+    if (str.substring(0, 4) != "type" && str.substring(0, 7) != "subtype" &&
+      /^\s*([\w]+)\s*(([\w]+)|([\w]+\(\s*([\d]+)(,\s*[\d]+)?\s*\)))((\s+)|;)((default|:=)(\s+)((\'[\w]*\')|[\d]+|[\w]+((\s+)[\w]+)?))?/i.exec(str)) {
```
what is this regexp doing? This seems overly complex when compared to the other ones.

RamonEsteveCuevas replied:
The reason why the regex is so large is because there are multiple ways to declare a variable. For instance we have the following ways to declare a variable:

l_number number;
l_number number(100);
l_number number(10, 2);
l_number number(10, 2) default 100.10;
l_string varchar2(100) default 'test variable';
l_object_type test_object := new test_object();

I'm not very good at writing regex so there is probably a much better way to find the variables inside the package headers. Currently im also not checking if type or subtype is not at the beginning (in the first part of the if statement).

martindsouza commented on lib/pmd.js:
```diff
@@ -57,20 +59,28 @@ pmd.validatePathRef = function(fullPath, objName){
   content.entities = []; //Holds list of entities for the object
 
-  content.json.forEach(function(jsonData){
+
```
Do an update on this file. This portion was fixed with PR 44

RamonEsteveCuevas replied:
Merged master in this fork and it is still working.

martindsouza commented on lib/pmd.js:
```diff
@@ -179,6 +203,34 @@ pmd.processFile = function(file){
+    if (entity.code && entity.type === pmd.DOCTYPES.EXCEPTIONS){
+      entity.exceptions = jsonData.ctx.exceptions;
+
+      // Loop over constants to see if there's one for this constantName
```
exceptions (in the comments, not constants)

RamonEsteveCuevas replied:
Changed it over, maybe there is a way to use an Object Oriented way to process the variables, constants and exceptions because they all use the same logic. Only some things are different (regex and the array where to append it to).

martindsouza commented:
@RamonEsteveCuevas I've made some comments in the PR.

As for the folder restructure I'm still unsure what problem this is solving. Is it that it's currently not searching through subfolders or it is? What do you want as final result? Do regular expressions resolve any of it?

RamonEsteveCuevas commented:
Currently it is not searching in subfolders (or I didn't find a way to do it). We have a project with more than 100 different amount of Packages/Object Types etc. These are all in different folders and subfolders. Currently you have to define all the folders and subfolders as entries in the config.json which is alot of work.

martindsouza commented:
@RamonEsteveCuevas I saw that you've responded to my previous feedback. I've been unable to review all the changes and hope to get to them soon. Thanks for applying them.

jgebal commented:
@martindsouza This PR looks very promising.
We're looking using your project to document utPLSQL version 3.
Would you be able to review this PR?

Would it be ok to reach out to you with any issues we encounter while using your documentation framework?

martindsouza commented:
@jgebal I'm glad to hear that you'll be using this for utPLSQL. For now I think we'll look at accepted the multiple folder searching and the `@exception` tag. I acknowledge that you'd like the `@variable` but still don't see value for user documentation.

RamonEsteveCuevas commented:
@martindsouza Made some changes in this pull request. We are using this framework now to document our own code and found some things we really needed. I have added these changes to the last commit. This commit contains the following:   
Merged the master into this fork;
Added support for package/type/view level description. These descriptions will be printed on top of the page. The message “No global description.” will be printed when there is no documentation found for the whole package/type/view;
Fixed the issue where the TOC was not printed when opening a package/view/type documentation;
Added support for commenting cursors inside package headers. These are now included in the API docs and have their own category (cursor) like procedures/function;
Changed the template so they include the package level description and the variables/exceptions;

vincentmorneau commented:
@RamonEsteveCuevas Great fixes. Looking forward to the merge, but I think you should take the html output files out of the PR.

RamonEsteveCuevas commented:
@vincentmorneau Will exclude them in the end commit. Put them inside the branch so you can easily see the results.

RamonEsteveCuevas commented:
@vincentmorneau Removed the output folder from the repository and added inside the gitignore. Is there any indication when this pull request will be merged, so we can start using the full version of this project?
